sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -106,6 +106,9 @@ class QueryExecution(
   lazy val toRdd: RDD[InternalRow] = new SQLExecutionRDD(
     executedPlan.execute(), sparkSession.sessionState.conf)
 
+  /** Get the metrics observed during the execution of the query plan. */
+  def observedMetrics: Map[String, Row] = CollectMetricsExec.collect(executedPlan)
```

Comment:
Based on this PR description, `observedMetrics ` is suggested as  a public API to use. 

However, it conflicts with what we commented above [in toRDD]. QueryExecution is not a public class and we discourage users to call the internal functions.

Reply:
I think `queryExecution` is exposed in `Dataset` as an unstable API. I think the comments in the class implies that as well:

>  \* The primary workflow for executing relational queries using Spark.  Designed to allow easy
>  \* access to the intermediate phases of query execution for developers.
>  \*
>  \* While this is not a public class, we should avoid changing the function names for the sake of
>  \* changing them, because a lot of developers use the feature for debugging.

I agree that using methods in this class here is discouraged though. Maybe we had to mark `Dataset.observe` as an unstable API or developer API too for now if it is difficult to avoid adding and using an API here.

Reply:
Technically, QueryExecution should be public class, as two methods in QueryExecutionListener are marked as `@DeveloperApi` and the signature of methods contain QueryExecution. If we worry about restriction for modifying QueryException due to being public class, we may need to have other class only for reporting to QueryExecutionListener.

Reply:
`@DeveloperApi` implies:

> A lower-level, unstable API intended for developers.

I agree that It is considered a public class but unstable. I am sure most of methods there are discouraged even for developers to use without knowing exactly what it does.

Reply:
Ah my bad. Thanks for pointing it out. 

Though I still feel listener stuff is more alike public API, as it can be used more than debug purpose. QueryExecutionListener and StreamingQueryListener take different approach, QueryExecutionListener directly exposes low-level internal instance and let end users use with caution, whereas StreamingQueryListener has individual classes which only contain information for report only, not exposing internal one.

IMHO, I feel StreamingQueryListener is the right way to go, but QueryExecutionListener has been served long ago, so no strong opinion.

Reply:
Yeah, `StreamingQueryProgress.observedMetrics` with `StreamingQueryListener` seems fine. The problem here looks only `QueryExecution.observedMetrics` with `QueryExecutionListener`, which looks having a contradiction about its stability. It seems it has to be fixed by either avoid adding it to `QueryExecution` or explicitly marking `Dataset.observe` as an unstable API (or developer API).

Reply:
+1. We should make the batch listener not rely on `QueryExecution`.

Reply:
I am not sure what the issue is?

By definition anything in `QueryExecution` is internal, unstable, API. The reason that I added it here is that is a good narrow waist to add this, you want to collect the metrics for the entire query. It is public because most other methods in this class are public; this allows for some clever integrations (for the more adventurous developer) and makes debugging easier.

The batch listener is marked as experimental. A developer should be warned when (s)he uses this API for anything (including using it to collect observable metrics). I am not sure how realistic stabilizing the batch listener API is if you include `QueryExecution` (or a stabilized version of it). We could expose a stable callback, e,g. `onObservedMetrics(...)`, for observed metrics. On the other hand is kind of annoying that we can't expose this through the Dataframe itself, and that is because when we execute the Dataframe we often use a different one under the hood.

Reply:
The issue is not about this PR itself, but about the bath query listener relying on `QueryExecution` which is not going to be public. I think the streaming query listener is better. It's still unstable now but we are going to make it stable in the future.

Reply:
The batch listener is no longer marked as experimental in 3.0, as there have been some of slight modifications (enough for `@Evolving`) but it doesn't change majorly during 4 years of life - see #25558. If we feel concerned it could be rolled back to Experimental/Unstable, though I know there're couple of projects in Spark ecosystem already leveraging it, and it represents that the possibility is not restricted to the debug purpose.

The major feature of batch listener is that (unlikely of streaming query listener) it exposes various plans for both logical/physical, and I don't imagine these plans will be used for execution-purposes. Mostly read-only. It may not unrealistic if we could provide these plans for read-only (cloned, can't execute), but yeah, it may not be easy as it seems (not enough familiar with it).